File: cli/lib/kb/sessions/adapters/claude_code.ex
defmodule KB.Sessions.Adapters.ClaudeCode do
  @moduledoc """
  Adapter for Claude Code session transcripts.

  Claude Code writes one JSONL per session to:

      ~/.claude/projects/<encoded-project-dir>/<session-uuid>.jsonl

  Encoded project dir names are the absolute cwd with `/` replaced by `-`,
  e.g. `-Users-naledi-Projects-hypha`. We strip to the last segment as the
  project slug.

  Sub-agent runs may live under `.../subagents/agent-*.jsonl` — we include
  them but mark them in metadata.
  """

  @behaviour KB.Sessions.Adapter

  alias KB.Sessions.{Redactor, SessionDoc}

  @store Path.join([System.user_home!() || "", ".claude", "projects"])

  @impl true
  def agent_name, do: "claude_code"

  @impl true
  def detect do
    if File.dir?(@store) do
      files =
        @store
        |> Path.join("**/*.jsonl")
        |> Path.wildcard()
        |> Enum.sort()

      {:ok, files}
    else
      :not_installed
    end
  end

  @impl true
  def parse(path) do
    with {:ok, raw} <- File.read(path) do
      records =
        raw
        |> String.split("\n", trim: true)
        |> Enum.flat_map(&decode_line/1)

      session_id = find_first(records, "sessionId") || uuid_from_path(path)
      cwd = find_first(records, "cwd") || ""
      started_at = find_first(records, "timestamp")
      model = find_model(records)

      messages = Enum.flat_map(records, &record_to_messages/1)

      doc = %SessionDoc{
        agent: agent_name(),
        project: project_slug_from_path(path, cwd),
        session_id: session_id,
        source_path: path,
        started_at: started_at,
        model: model,
        messages: messages,
        metadata: %{
          subagent: subagent?(path),
          record_count: length(records)
        }
      }

      {:ok, doc}
    end
  end

  @impl true
  def redact(%SessionDoc{} = doc) do
    opts = redact_opts()

    redacted_messages =
      Enum.map(doc.messages, fn m ->
        Map.update!(m, :content, &Redactor.redact(&1, opts))
      end)

    %SessionDoc{doc | messages: redacted_messages}
  end

  # ─── helpers ──────────────────────────────────────────────────────────

  defp decode_line(""), do: []

  defp decode_line(line) do
    case Jason.decode(line) do
      {:ok, %{} = record} -> [record]
      _ -> []
    end
  end

  defp find_first(records, key) do
    Enum.find_value(records, fn r -> Map.get(r, key) end)
  end

  defp find_model(records) do
    Enum.find_value(records, fn
      %{"message" => %{"model" => model}} when is_binary(model) -> model
      _ -> nil
    end)
  end

  # Claude Code emits user / assistant records with a `message` field
  # shaped like `%{"role" => "user", "content" => string | [blocks]}`.
  defp record_to_messages(%{"type" => "user", "message" => %{"content" => content} = msg} = r) do
    text = extract_text(content)

    if text == "" do
      []
    else
      [%{role: "user", content: text, timestamp: r["timestamp"], model: msg["model"]}]
    end
  end

  defp record_to_messages(
         %{"type" => "assistant", "message" => %{"content" => content} = msg} = r
       ) do
    text = extract_text(content)

    if text == "" do
      []
    else
      [%{role: "assistant", content: text, timestamp: r["timestamp"], model: msg["model"]}]
    end
  end

  # Tool results / side-channel records — skip by default.
  defp record_to_messages(_), do: []

  defp extract_text(content) when is_binary(content), do: String.trim(content)

  defp extract_text(content) when is_list(content) do
    content
    |> Enum.map(&block_to_text/1)
    |> Enum.reject(&(&1 == ""))
    |> Enum.join("\n\n")
    |> String.trim()
  end

  defp extract_text(_), do: ""

  defp block_to_text(%{"type" => "text", "text" => t}) when is_binary(t), do: t
  defp block_to_text(%{"type" => "thinking"}), do: ""

  defp block_to_text(%{"type" => "tool_use", "name" => name, "input" => input}) do
    "`[tool: #{name}]` `#{Jason.encode!(input)}`"
  end

  defp block_to_text(%{"type" => "tool_result", "content" => c}) when is_binary(c) do
    "```tool-result\n#{c}\n```"
  end

  defp block_to_text(%{"type" => "tool_result", "content" => c}) when is_list(c) do
    extract_text(c)
  end

  defp block_to_text(%{"text" => t}) when is_binary(t), do: t
  defp block_to_text(_), do: ""

  defp project_slug_from_path(path, cwd) do
    cond do
      is_binary(cwd) and cwd != "" -> cwd |> Path.basename() |> slugify()
      true -> path |> Path.dirname() |> Path.basename() |> decode_project_dir() |> slugify()
    end
  end

  # `-Users-naledi-Projects-hypha` -> `hypha`
  defp decode_project_dir("-" <> rest), do: rest |> String.split("-") |> List.last() |> to_string()
  defp decode_project_dir(other), do: other

  defp slugify(nil), do: "unknown-project"
  defp slugify(""), do: "unknown-project"

  defp slugify(s) do
    s
    |> String.downcase()
    |> String.replace(~r/[^a-z0-9]+/, "-")
    |> String.trim("-")
    |> case do
      "" -> "unknown-project"
      slug -> slug
    end
  end

  defp uuid_from_path(path), do: path |> Path.basename() |> Path.rootname()

  defp subagent?(path) do
    String.contains?(path, "subagents/") or String.starts_with?(Path.basename(path), "agent-")
  end

  defp redact_opts do
    case File.read("kb.config.json") do
      {:ok, json} ->
        case Jason.decode(json) do
          {:ok, config} ->
            extras = get_in(config, ["ingest", "redaction", "extra_patterns"]) || []
            [extra_patterns: extras]

          _ ->
            []
        end

      _ ->
        []
    end
  end
end

File: packages/adapters/src/sessions/claude-code.ts
import { existsSync, readFileSync, readdirSync, statSync } from 'node:fs';
import { basename, dirname, extname, join } from 'node:path';
import { homedir } from 'node:os';
import { redact } from './redactor.ts';
import type {
  DetectResult,
  ParseResult,
  SessionAdapter,
  SessionDoc,
  SessionMessage,
  RedactOpts,
} from './types.ts';

const STORE = join(homedir(), '.claude', 'projects');

function walkJsonl(dir: string): string[] {
  const out: string[] = [];
  let entries: import('node:fs').Dirent[];
  try {
    entries = readdirSync(dir, { withFileTypes: true });
  } catch {
    return out;
  }
  for (const e of entries) {
    const full = join(dir, e.name);
    if (e.isDirectory()) {
      out.push(...walkJsonl(full));
    } else if (e.isFile() && full.endsWith('.jsonl')) {
      out.push(full);
    }
  }
  return out;
}

export const claudeCodeAdapter: SessionAdapter = {
  agentName: 'claude_code',

  detect(): DetectResult {
    if (!existsSync(STORE)) return { kind: 'not_installed' };
    return { kind: 'ok', files: walkJsonl(STORE).sort() };
  },

  parse(path: string): ParseResult {
    let raw: string;
    try {
      raw = readFileSync(path, 'utf8');
    } catch (e) {
      return { ok: false, reason: String(e) };
    }
    const records: Record<string, unknown>[] = [];
    for (const line of raw.split('\n')) {
      if (!line.trim()) continue;
      try {
        const parsed = JSON.parse(line);
        if (parsed && typeof parsed === 'object') records.push(parsed);
      } catch {
        // skip malformed lines
      }
    }

    const sessionId =
      (findFirst(records, 'sessionId') as string | undefined) ?? uuidFromPath(path);
    const cwd = (findFirst(records, 'cwd') as string | undefined) ?? '';
    const startedAt = findFirst(records, 'timestamp') as string | undefined;
    const model = findModel(records);
    const messages = records.flatMap(recordToMessages);

    return {
      ok: true,
      doc: {
        agent: 'claude_code',
        project: projectSlugFromPath(path, cwd),
        sessionId,
        sourcePath: path,
        startedAt,
        model,
        messages,
        metadata: {
          subagent: isSubagent(path),
          recordCount: records.length,
        },
      },
    };
  },

  redact(doc: SessionDoc, opts: RedactOpts = {}): SessionDoc {
    return {
      ...doc,
      messages: doc.messages.map((m) => ({ ...m, content: redact(m.content, opts) })),
    };
  },
};

function findFirst(records: Record<string, unknown>[], key: string): unknown {
  for (const r of records) {
    if (r[key] != null) return r[key];
  }
  return undefined;
}

function findModel(records: Record<string, unknown>[]): string | undefined {
  for (const r of records) {
    const msg = r['message'];
    if (msg && typeof msg === 'object' && 'model' in msg) {
      const m = (msg as { model?: unknown }).model;
      if (typeof m === 'string') return m;
    }
  }
  return undefined;
}

function recordToMessages(r: Record<string, unknown>): SessionMessage[] {
  const type = r['type'];
  const msg = r['message'];
  if (type !== 'user' && type !== 'assistant') return [];
  if (!msg || typeof msg !== 'object') return [];
  const message = msg as { content?: unknown; model?: unknown };
  const text = extractText(message.content);
  if (!text) return [];
  return [
    {
      role: type as 'user' | 'assistant',
      content: text,
      timestamp: typeof r['timestamp'] === 'string' ? (r['timestamp'] as string) : undefined,
      model: typeof message.model === 'string' ? message.model : undefined,
    },
  ];
}

function extractText(content: unknown): string {
  if (typeof content === 'string') return content.trim();
  if (Array.isArray(content)) {
    return content.map(blockToText).filter((s) => s !== '').join('\n\n').trim();
  }
  return '';
}

function blockToText(block: unknown): string {
  if (!block || typeof block !== 'object') return '';
  const b = block as Record<string, unknown>;
  const type = b.type;
  if (type === 'text' && typeof b.text === 'string') return b.text;
  if (type === 'thinking') return '';
  if (type === 'tool_use' && typeof b.name === 'string') {
    try {
      return `\`[tool: ${b.name}]\` \`${JSON.stringify(b.input ?? {})}\``;
    } catch {
      return `\`[tool: ${b.name}]\``;
    }
  }
  if (type === 'tool_result') {
    if (typeof b.content === 'string') return '```tool-result\n' + b.content + '\n```';
    if (Array.isArray(b.content)) return extractText(b.content);
  }
  if (typeof b.text === 'string') return b.text;
  return '';
}

function projectSlugFromPath(path: string, cwd: string): string {
  if (cwd) return slugify(basename(cwd));
  return slugify(decodeProjectDir(basename(dirname(path))));
}

function decodeProjectDir(name: string): string {
  if (name.startsWith('-')) {
    const parts = name.slice(1).split('-');
    return parts[parts.length - 1] ?? '';
  }
  return name;
}

function slugify(s: string | null | undefined): string {
  if (!s) return 'unknown-project';
  const slug = s.toLowerCase().replace(/[^a-z0-9]+/g, '-').replace(/^-+|-+$/g, '');
  return slug || 'unknown-project';
}

function uuidFromPath(path: string): string {
  const base = basename(path);
  return base.slice(0, base.length - extname(base).length);
}

function isSubagent(path: string): boolean {
  return path.includes('subagents/') || basename(path).startsWith('agent-');
}

// keep statSync import live for tree-shaking-resistant typing
void statSync;
